File: src/api/env.ts
import { invokeCommand } from "./client";

/** 与 Rust env_service DTO 对齐（camelCase） */

export interface PythonEnv {
  configured: boolean;
  ready: boolean;
  path?: string | null;
  version?: string | null;
  hint?: string | null;
}

export interface NodeEnv {
  ready: boolean;
  path?: string | null;
  version?: string | null;
  npmGlobalRoot?: string | null;
  hint?: string | null;
}

export interface FridaEnv {
  pythonReady: boolean;
  installed: boolean;
  fridaVersion?: string | null;
  fridaToolsVersion?: string | null;
  hint?: string | null;
}

export interface McpEnv {
  reachable: boolean;
  port: number;
  hint?: string | null;
  /** 宿主应用是否存在（mac 检测；其他平台暂为 null） */
  appInstalled?: boolean | null;
  /** 检测到的应用路径（mac） */
  appPath?: string | null;
}

export interface ProcPath {
  name: string;
  path: string;
  summary?: string | null;
  readable: boolean;
}

export interface ForegroundApp {
  /** ready | adb_unavailable | no_device | no_foreground | error */
  state: string;
  serial?: string | null;
  package?: string | null;
  activity?: string | null;
  pid?: string | null;
  nativeLibDir?: string | null;
  procPaths: ProcPath[];
  hint?: string | null;
  error?: string | null;
}

/** env_resolve_interpreter 返回：picked→resolved 解析结果 */
export interface ResolvedInterpreter {
  pickedPath: string;
  resolvedPath: string;
  version?: string | null;
  /** as-is | pyenv-shim | app-bundle | symlink */
  how: string;
}

export interface EnvOverview {
  python: PythonEnv;
  node: NodeEnv;
  frida: FridaEnv;
  idaMcp: McpEnv;
  jadxMcp: McpEnv;
}

export const envApi = {
  python(): Promise<PythonEnv> {
    return invokeCommand<PythonEnv>("env_python");
  },
  node(): Promise<NodeEnv> {
    return invokeCommand<NodeEnv>("env_node");
  },
  /** Frida 检测依赖 Python 配置，前端在 Python 就绪后才启用该查询（剪枝） */
  frida(): Promise<FridaEnv> {
    return invokeCommand<FridaEnv>("env_frida");
  },
  idaMcp(): Promise<McpEnv> {
    return invokeCommand<McpEnv>("env_ida_mcp");
  },
  jadxMcp(): Promise<McpEnv> {
    return invokeCommand<McpEnv>("env_jadx_mcp");
  },
  /** 安卓前台应用；serial 缺省自动选第一台在线设备；adb 不可用返回 adb_unavailable */
  foreground(serial?: string): Promise<ForegroundApp> {
    return invokeCommand<ForegroundApp>("env_foreground", {
      args: { serial: serial ?? null },
    });
  },
  overview(): Promise<EnvOverview> {
    return invokeCommand<EnvOverview>("env_overview");
  },
  /** 文件选择器选中项 → 真解释器路径（pyenv shim/.app/symlink 解析） */
  resolveInterpreter(picked: string): Promise<ResolvedInterpreter> {
    return invokeCommand<ResolvedInterpreter>("env_resolve_interpreter", { picked });
  },
  /** python 选择器建议起始目录 */
  pythonStartDir(): Promise<string> {
    return invokeCommand<string>("env_python_start_dir");
  },
};

File: src-tauri/src/commands/env.rs
//! 环境探测命令层（P7）：各环境卡独立刷新 + 聚合 overview。
//! 只做转发；探测、剪枝、解析全在 EnvService。
//! ADB 卡继续复用 commands/device.rs 的 adb_environment（同一 runner）。

use crate::AppState;
use crate::core::error::CoreResult;
use crate::services::env_service::{
    EnvOverview, ForegroundApp, FridaEnv, McpEnv, NodeEnv, PythonEnv,
};

#[tauri::command]
pub async fn env_python(state: tauri::State<'_, AppState>) -> CoreResult<PythonEnv> {
    Ok(state.env.python().await)
}

#[tauri::command]
pub async fn env_node(state: tauri::State<'_, AppState>) -> CoreResult<NodeEnv> {
    Ok(state.env.node().await)
}

#[tauri::command]
pub async fn env_frida(state: tauri::State<'_, AppState>) -> CoreResult<FridaEnv> {
    Ok(state.env.frida().await)
}

#[tauri::command]
pub async fn env_ida_mcp(state: tauri::State<'_, AppState>) -> CoreResult<McpEnv> {
    Ok(state.env.ida_mcp().await)
}

#[tauri::command]
pub async fn env_jadx_mcp(state: tauri::State<'_, AppState>) -> CoreResult<McpEnv> {
    Ok(state.env.jadx_mcp().await)
}

/// 安卓前台应用：按设备查询（P8 设备页）；serial 缺省自动选第一台在线设备
#[tauri::command]
pub async fn env_foreground(
    state: tauri::State<'_, AppState>,
    args: Option<ForegroundArgs>,
) -> CoreResult<ForegroundApp> {
    let serial = args.and_then(|a| a.serial).filter(|s| !s.is_empty());
    Ok(state.env.foreground_on(serial).await)
}

#[derive(Debug, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ForegroundArgs {
    pub serial: Option<String>,
}

/// 仪表盘聚合：一次拿全部环境卡（前端各卡仍可独立刷新）
#[tauri::command]
pub async fn env_overview(state: tauri::State<'_, AppState>) -> CoreResult<EnvOverview> {
    Ok(state.env.overview().await)
}

/// 文件选择器解析：把选中的「错误入口」（pyenv shim / .app / symlink）解析成真解释器
#[tauri::command]
pub async fn env_resolve_interpreter(
    state: tauri::State<'_, AppState>,
    picked: String,
) -> CoreResult<crate::services::env_service::ResolvedInterpreter> {
    Ok(state.env.resolve_interpreter(&picked).await)
}

/// python 选择器建议起始目录（pyenv versions 优先）
#[tauri::command]
pub async fn env_python_start_dir(state: tauri::State<'_, AppState>) -> CoreResult<String> {
    Ok(state.env.python_start_dir().await)
}
